storage/host-flash/host-flash.c

```diff
@@ -15,155 +15,167 @@
+	return read;
 }
 
 
 ssize_t hostflash_write(unsigned int addr, void *buff, size_t bufflen)
 {
-    char currentByte = 0;
-    size_t wrote = 0;
-    ssize_t stat;
-
-    if (addr + bufflen > hostflash_common.flashsz)
-        return -EINVAL;
-
-    while (wrote != bufflen) {
-        stat = hostflash_read(addr + wrote, &currentByte, 1);
-        if (stat <= 0)
-            return stat;
-
-        currentByte = currentByte & *((char *)buff + wrote);
-        stat = pwrite(hostflash_common.filefd, &currentByte, 1, addr + wrote);
-        if (stat < 0) {
-            if (errno == -EINTR)
-                continue;
-
-            return stat;
-        }
-
-        wrote++;
-    }
-
-    return (ssize_t)wrote;
+	char tempTab[256];
+	size_t wrote = 0, i;
+	int len;
+	ssize_t stat, readed;
+
+	if (addr + bufflen > hostflash_common.flashsz)
+		return -EINVAL;
+
+	while (wrote != bufflen) {
+		if ((bufflen - wrote) / sizeof(tempTab))
+			len = sizeof(tempTab);
+		else
+			len = (bufflen - wrote) % sizeof(tempTab);
+
+		readed = hostflash_read(addr + wrote, tempTab, len);
+		if (readed <= 0)
+			return readed;
+
+		for (i = 0; i < readed; i++)
+			tempTab[i] = tempTab[i] & *((char *)buff + wrote + i);
+
+		len = 0;
+		while (len != readed) {
+			stat = pwrite(hostflash_common.filefd, &tempTab, readed - len, addr + wrote + len);
+			if (stat < 0) {
+				if (errno == EINTR)
+					continue;
+
+				return stat;
+			}
+			len += stat;
+		}
+		wrote += len;
+	}
+
+	return (ssize_t)wrote;
 }
 
 
 void hostflash_sectorErase(unsigned int addr)
 {
-    char tempTab[256];
-    ssize_t len = sizeof(tempTab);
-    int erased = 0;
-    unsigned int sectorNum;
-    unsigned int sectorAddr;
-    int stat;
-
-    memset(tempTab, 0xFF, sizeof(tempTab));
-
-    if (addr < hostflash_common.flashsz) {
-        sectorNum = addr / hostflash_common.sectorsz;
-        sectorAddr = sectorNum * hostflash_common.sectorsz;
-    }
-    else {
-        return;
-    }
-
-    while (erased != hostflash_common.sectorsz) {
-        stat = pwrite(hostflash_common.filefd, tempTab, len, sectorAddr + erased);
-        if (stat < 0) {
-            if (errno == -EINTR)
-                continue;
-
-            return;
-        }
-        else {
-            len = len - stat;
-            if (len == 0)
-                len = sizeof(tempTab);
-        }
-
-        erased += stat;
-    }
-
-    return;
+	char tempTab[256];
+	ssize_t len = sizeof(tempTab);
+	int erased = 0;
+	unsigned int sectorNum;
+	unsigned int sectorAddr;
+	int stat;
+
+	memset(tempTab, 0xFF, sizeof(tempTab));
+
+	if (addr < hostflash_common.flashsz) {
+		sectorNum = addr / hostflash_common.sectorsz;
+		sectorAddr = sectorNum * hostflash_common.sectorsz;
+	}
+	else {
+		return;
+	}
+
+	while (erased != hostflash_common.sectorsz) {
+		stat = pwrite(hostflash_common.filefd, tempTab, len, sectorAddr + erased);
+		if (stat < 0) {
+			if (errno == EINTR)
+				continue;
+
+			return;
+		}
+		else {
+			len = len - stat;
```

Comment:
You can type this as `len -= stat;` but this a matter of taste, you can leave it is, if you prefer it.

Reply:
I prefer `len -= stat;` :) Done